File: ruby/bob/bob.rb
class Bob

  def hey(remark)
    if is_yelling?(remark)
      "Whoa, chill out!"
    elsif is_questioning?(remark)
      "Sure."
    elsif is_silent?(remark)
      "Fine. Be that way!"
    else
      "Whatever."
    end
  end


  def is_yelling?(remark)
    remark == remark.upcase && remark =~ /[A-Z]/
  end

  def is_questioning?(remark)
    remark[-1, 1] == "?"
  end

  def is_silent?(remark)
    !remark.match(/\S/)
  end


end

















# the little methods below didn't help.  I will now try it with boolean methods

# class Bob

#   def hey(remark)
#     return detect_yelling(remark)
#     detect_question(remark)
#     detect_silence(remark)
#     "Whatever."
#   end


#   def detect_yelling(remark)
#     if remark == remark.upcase && remark =~ /[A-Z]/
#       return "Whoa, chill out!"
#     end
#   end

#   def detect_question(remark)
#     if remark[-1, 1] == "?"
#       return "Sure."
#     end
#   end

#   def detect_silence(remark)
#     if !remark.match(/\S/) == true
#       return "Fine. Be that way!"
#     end
#   end


# end

File: javascript/bob/bob.js
var Bob = function() {};

Bob.prototype.hey = function(input) {

  var isYelling, isQuestioning, isSilent;

  var isYelling = function() {
    return input === input.toUpperCase() && input !== input.toLowerCase();
  };

  var isQuestioning = function() {
    return input.endsWith('?');
  };

  var isSilent = function() {
    return !input.match(/\S/);
  };


  if (isYelling() === true) {
    return "Whoa, chill out!";
  }
  else if (isQuestioning() === true) {
    return "Sure.";
  }
  else if (isSilent() === true) {
    return "Fine. Be that way!";
  }
  else {
    return "Whatever.";
  };
};

module.exports = Bob;


if (!String.prototype.endsWith) {
  String.prototype.endsWith = function(searchString, position) {
      var subjectString = this.toString();
      if (typeof position !== 'number' || !isFinite(position) || Math.floor(position) !== position || position > subjectString.length) {
        position = subjectString.length;
      }
      position -= searchString.length;
      var lastIndex = subjectString.indexOf(searchString, position);
      return lastIndex !== -1 && lastIndex === position;
  };
}